File: cpp/637.cpp
#include <iostream>
#include <vector>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  void helper(TreeNode *t, int depth, vector<double> &ret, vector<int> &cnt) {
    if (!t)
      return;
    if (depth > ret.size()) {
      ret.push_back(0);
      cnt.push_back(0);
    }
    ret[depth-1] = (ret[depth-1] * cnt[depth-1] + t->val) / (++cnt[depth]);
    depth++;
    helper(t->left, depth, ret, cnt);
    helper(t->right, depth, ret, cnt);
  }
  vector<double> averageOfLevels(TreeNode *root) {
    vector<double> ret;
    vector<int> count;
    helper(root, 1, ret, count);
    return ret;
  }
};

File: go/637.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func averageOfLevels(root *TreeNode) []float64 {
  // init
  var ret []float64 = make([]float64, 0)
  var q []*TreeNode = make([]*TreeNode, 0)

  q = append(q, root)
  var size int
  // BFS method
  for len(q) > 0 {
    size = len(q)
    ret = append(ret, 0)
    for i := 0 ; i < size ; i++ {
      ret[len(ret)-1] += float64(q[0].Val)
      if q[0].Left !=nil {
        q = append(q, q[0].Left)
      }
      if q[0].Right != nil {
        q = append(q, q[0].Right)
      }
      q = q[1:]
    }
    ret[len(ret)-1] /= float64(size)
  }
  return ret
}
